### PR TITLE
Hide Showcase Banners for Active Themes

### PR DESCRIPTION
**Pull request by mendezcode**
Make sure banners are not displayed for active themes.

#23122

### CODE REVIEW

**Comment by mendezcode** [comment]
I was able to fix it in 0d40543 by using the `isLoggedIn` prop.

Since the initial problem was that the banner didn't render accurately because at a point in time, `currentThemeId` was returning null from the state, even if `isLoggedIn` was true.

To mitigate this, we render `null` if the user is logged-in, but no `currentThemeId` is set. Otherwise, we render the filtered banners, omitting the one for the current theme:

```js
let themeBanners = null;

const availableBanners = {
	'photo-blog': PhotoBlogBanner,
	'small-business': SmallBusinessBanner,
};

if ( isLoggedIn ) {
	if ( this.props.siteId && currentThemeId ) {
		themeBanners = values( omit( availableBanners, currentThemeId ) );
	}
} else {
	themeBanners = values( availableBanners );
}
```
Then, on `render()`, we do this:

```jsx
{ themeBanners ? <RandomThemesBanner banners={ themeBanners } /> : null }
```

Works like a charm. Also tested SSR.

**Comment by ockham** [comment]
Okay, so the issue we're seeing is due to `<RandomThemeBanner />` picking the random banner to display only once -- in `componentDidMount`. Thus, it doesn't update upon a `banners` prop change caused by `currentThemeId` changing from `null` to the actual theme ID.

0d40543 actually boils down to not rendering `<RandomThemeBanner />` when `isLoggedIn && ! ( this.props.siteId && currentThemeId )` (in which case `themeBanners` won't be set, and  `<RandomThemeBanner />` thus won't be rendered). The `isLoggedIn` and `siteId` checks aren't really necessary tho; we can simplify to

```jsx
const themeBanners = omit(
	{
		'photo-blog': PhotoBlogBanner,
		'small-business': SmallBusinessBanner,
	},
	currentThemeId
); // We don't want to advertise the theme that's already active.

{ currentThemeId ? <RandomThemesBanner banners={ banners={ values( themeBanners ) } /> : null }
```

The last line could be simplified even more to
```jsx
{ currentThemeId && <RandomThemesBanner banners={ banners={ values( themeBanners ) } /> }
```

**Comment by ockham** [comment]
However, this still won't update the banner e.g. if we're activating the theme it is advertising. This is somewhat trickier; I think I have a vague idea how to tackle that case...

**Comment by ockham** [comment]
Pushed a few commits that should now take care of all edge cases, as shown by the video -- especially when activating a different theme:

* If the theme that the banner is advertising is activated, display a different banner.
* Otherwise, don't change the banner.

themebanner

**Comment by mendezcode** [comment]
@ockham Thanks a lot for the changes! Yes, the issue when activating the theme is something to have a deeper look into. I also had it in mind while developing the fix.

I've been testing your patch, and I noticed a small issue with it. The banner randomly rotates between the active banners, while the state is being populated. This can be seen here:

gif-screenshot

This was exactly the reason why I was checking for `isLoggedIn`, `siteID`, and `currentThemeId`, since when those three values are available and set, we can safely render the banner and the randomly chosen banner will be consistently rendered.

I'm working with the code you added to see how can we make it work, as well as the activation fix too.

Please let me know your thoughts on the above!

**Comment by ockham** [comment]
I think the only reasonable way around that issue is to add back the `{ currentThemeId && ... }` check. Before we get a hold of `currentThemeId`, we plainly can't reliably display a random banner and assume it's not advertising the current theme. (I'm pretty sure `currentThemeId` is sufficient, the other criteria are more circumstantial.)

**Comment by ockham** [comment]
It's correct that we only get a `currentThemeId` in single-site mode, so you're right that we'll need more criteria to make sure we _are_ showing the banner in multi-site and logged-out modes. We shouldn't modify the `banners` const tho, all we need and want is a bool for the component's visibility.

Theoretically, something like

```js
const showBanner = currentThemeId || ! siteId || ! isLoggedIn;
/*...*/
{ showBanner && <RandomThemesBanner /*...*/ />
```

should cover the single-site, multi-site, and logged-out cases. In practice however, you're still likely to see the little flicker you're seeing now, since while the showcase is loading, both `currentThemeId` and `siteId` are probably going to be falsey. So you'll also need to add a check whether we're currently requesting sites (we have `isRequestingSite` and `isRequestingAllSites` selectors for that IIRC, you'll need both), and possibly even something similar for `isLoggedIn` (i.e. `isRequestingUser`, in case something like that exists).

File: client/my-sites/themes/theme-showcase.jsx [hunk 1]
```diff
@@ -204,13 +204,16 @@ class ThemeShowcase extends React.Component {
 				.sort( ( a, b ) => a.order - b.order )
 		);
 
+		const showBanners = currentThemeId || ! siteId || ! isLoggedIn;
+
 		const themeBanners = omit(
+			// We don't want to advertise the theme that's already active.
```
Comment (ockham) [inline client/my-sites/themes/theme-showcase.jsx]: Nitpick - it's a bit nasty to have the comment _inside_ the function call. Maybe move it up one line?

Comment (mendezcode) [reply client/my-sites/themes/theme-showcase.jsx]: Ha! That comes from Prettier, moving...

**Comment by mendezcode** [comment]
Awesome! :highfive:

**Pull request merged**